specs/console-output.md

```diff
@@ -0,0 +1,153 @@
+`Tip`|`Blue`|<span style="color: blue">(i)</span>
+`Failed`|`Red`|<span style="color: red">! →</span>
+`Chaos`|`DarkRed`|<span style="color: darkred">× →</span>
+`Mocked`|`DarkYellow`|<span style="color: orange">o →</span>
+
+![Sample output in a terminal](./assets/logging-boxed-ascii-icons.png)
+
+#### Nerd Font icons colors
+
+When using Nerd Font icons, both icons and messages are colored.
+
+Following is the list of colors for each label:
+
+Message type|Foreground color|Example
+------------|----------------|-------
+`Normal`|`Default`|
+`InterceptedRequest`|`Default`|<span style="font-family: 'Hack Nerd Font'">&#62529;</span>&nbsp;&nbsp;&nbsp;(f441)
+`PassedThrough`|`Gray`|<span style="font-family: 'Hack Nerd Font'; color: grey">&#58172;</span>&nbsp;&nbsp;&nbsp;(e33c)
+`Warning`|`Yellow`|<span style="font-family: 'Hack Nerd Font'; color: yellow">&#62497;</span>&nbsp;&nbsp;&nbsp;(f421)
+`Tip`|`Blue`|<span style="font-family: 'Hack Nerd Font'; color: blue">&#64486;</span>&nbsp;&nbsp;&nbsp;(fbe6)
+`Failed`|`Red`|<span style="font-family: 'Hack Nerd Font'; color: red">&#63067;</span>&nbsp;&nbsp;&nbsp;(f65b)
+`Chaos`|`DarkRed`|<span style="font-family: 'Hack Nerd Font'; color: darkred">&#61832;</span>&nbsp;&nbsp;&nbsp;(f188)
+`Mocked`|`DarkYellow`|<span style="font-family: 'Hack Nerd Font'; color: orange">&#61540;</span>&nbsp;&nbsp;&nbsp;(f064)
+
+![Sample output in a terminal](./assets/logging-boxed-nerdfont-icons.png)
+
+### Message context
+
+Because messages can be printed out of band (ie. multiple messages about intercepted requests, followed by Proxy's response to each request), it's important to know which message is related to which request. To help with this, each message is suffixed with the request's method and URL.
+
+```text
+↑ ↑  ┌ passed through
+     └ GET https://graph.microsoft.com/v1.0/me/joinedTeams
+```
+
+### Multiline messages
+
+To improve readability, multiline messages will be visually grouped together, using boxes. The first and the last line of the message will be indicated with respectively top-left-, and bottom-left corner symbol. Middle lines of messages of 3 and more lines will be indicated with a | (pipe) symbol.
+
+```text
+↑ ↑  ┌ passed through
+     └ GET https://graph.microsoft.com/v1.0/me/joinedTeams
+```
+
+```text
+/!\  ┌ To improve performance of your application, use the $select parameter.
+     │ More info at https://learn.microsoft.com/graph/query-parameters#select-parameter
+     └ GET https://graph.microsoft.com/v1.0/me/joinedTeams
+```
+
+## Logging method
+
+To support requirements listed above, and ensure consistent logging from the Proxy's core and its plugins, we'll implement a centralized method, that accepts an array of messages, the type of the message, and context information about the request:
+
+```cs
+interface RequestContext {
+    public ProxyRequestArgs RequestArgs { get; }
+}
+
+Log(string[] messages, MessageType messageType, RequestContext: context): void
```

Comment:
Do we want to make the Context parameter optional to handle the case where we don't have a SessionEventArgs object to pass?

Reply:
Good point! In that case we'll have a log message that's not related to any request. Not something we use at the moment, but could be useful in the future to avoid breaking the interface.